### ISSUE TITLE
Parser shouldn't assume that every class, mixin, and extension has a body

### ISSUE THREAD

**Issue by bwilkerson** [dart-lang/sdk#53256]
The parser contains an assumption that all classes, mixins, and extensions will have a body, but this isn't true in code that is incomplete. In particular, the assumption occurs here: https://github.com/dart-lang/sdk/blob/main/pkg/_fe_analyzer_shared/lib/src/parser/parser_impl.dart#L4243. We assert that the next token is `{` without checking.

This is causing poor recovery in production and an exception in tests that prevents us from testing the behavior seen in production.

I'd like to propose that we verify the existence of a body and provide a synthetic body if there isn't one.

**Comment by bwilkerson** [comment]
@jensjoha @johnniwinther

**Comment by bwilkerson** [comment]
Clarification. After looking more carefully I realized that it isn't the method I pointed at that's at fault. Because of the listener style interface recovery has to happen at the call site, and that isn't happening for extension types. Recovery needs to be added just before https://github.com/dart-lang/sdk/blob/main/pkg/_fe_analyzer_shared/lib/src/parser/parser_impl.dart#L3120.

As a test case, I was using
```dart
extension type E(int i)
```
(trying to test completion after the representation declaration).
